feat(da-throttling): Introduce DA throttling via the `MinerAPI`

0xKitsune opened the pull request:
This PR introduces DA throttling to `rollup-boost` as well as fixes RPC module initialization. Note that this PR builds on #61, incorporating its changes.



### Closes #48

This PR introduces DA throttling by implementing the [MinerApiExtServer](https://docs.rs/op-alloy-rpc-jsonrpsee/0.9.6/op_alloy_rpc_jsonrpsee/traits/trait.MinerApiExtServer.html) as an RPC module. Additionally, a `MinerApi` trait has been implemented for the remaining `miner_` RPC endpoints. Tests were added for all newly introduced RPC modules.

While Reth provides a [MinerApi](https://github.com/paradigmxyz/reth/blob/e2c45f2b1fd5a0c50f2440175a61e23410ca06ee/crates/rpc/rpc/src/miner.rs#L14-L24), it is synchronous where `rollup-boost` requires an async implementation. To address this we recreated the trait within `rollup-boost` and can upstream this to Reth, allowing us to switch to `reth-rpc-api` once an async version is available.

Since the `auth_rpc` does not expose the Miner API, we introduced a new `ExecutionClient` struct to orchestrate requests to either the authenticated Engine API (via `auth_client`) or the JSON-RPC API (via `client`).  

```rust
/// Client interface for interacting with an execution layer node.
///
/// - **Engine API** calls are faciliated via the `auth_client` (requires JWT authentication).
/// -  All other API calls including the **Eth & Miner APIs** are faciliated via the `client` (optional JWT authentication).
///
#[derive(Clone)]
pub struct ExecutionClient<C: ClientT = HttpClient<HttpBackend>> {
    /// Handles requests to Eth, Miner, and other execution layer APIs (optional JWT authentication)
    pub client: Arc<C>,
    /// Address of the RPC server for execution layer API calls, excluding the Engine API
    pub http_socket: SocketAddr,
    /// Handles requests to the authenticated Engine API (requires JWT authentication)
    pub auth_client: Arc<HttpClient<AuthClientService<HttpBackend>>>,
    /// Address of the RPC server for authenticated Engine API calls
    pub auth_socket: SocketAddr,
}
```

### Fixes #63

Previously,the only RPC methods initialized were from the Engine API. This has been updated to ensure that `EthApiServer`, `MinerApiServer`, and `MinerApiExtServer` are properly merged into the RPC module. 

```rust
// --snip--
fn try_into(self) -> Result<RpcModule<()>, Self::Error> {
    let mut module: RpcModule<()> = RpcModule::new(());
    module.merge(EngineApiServer::into_rpc(self.clone()))?;
    module.merge(EthApiServer::into_rpc(self.clone()))?;
    module.merge(MinerApiServer::into_rpc(self.clone()))?;
    module.merge(MinerApiExtServer::into_rpc(self))?;

    Ok(module)
}
```


### Addresses #65
Since we introduced a `client` and `auth_client` for each execution client, we made a small update to the proxy logic to ensure that all non-multiplexed methods are routed to the auth RPC of the default L2 client with the proper authorization header.

avalonche commented on src/proxy.rs:
```diff
@@ -88,18 +94,24 @@ where
-                *req.uri_mut() = target_url;
+                *req.uri_mut() = target_url.clone();
+                // Insert JWT Authorization headers
+                req.headers_mut()
```
why is there a need to insert the auth header here when it can be cloned from the original op-node request?

0xOsiris replied:
Because `op-node` is not the only service pointing at rollup-boost. `op-batcher` for example will be sending `eth_getBlockNumber` to the `http.api` of `op-geth`, but we are pointed at the Auth API, therefore we need to override the headers.

avalonche reviewed (commented):
could you update the .env.example and README with the new cli commands?

avalonche commented on src/proxy.rs:
```diff
@@ -88,18 +94,24 @@ where
 
             debug!(
```
this log can be remove since there is info logging of the rpc method

avalonche commented on src/main.rs:
```diff
@@ -164,91 +131,67 @@ async fn main() -> Result<()> {
+    } else if let Some(path) = builder_args.builder_auth_jwtsecret_path.as_ref() {
+        JwtSecret::from_file(path)?
+    } else {
+        bail!("Missing Builder JWT secret");
```
this would make the builder_jwt mandatory. Could it fallback to the l2 jwt token?

0xOsiris replied:
I don't have a strong opinion here, but it seems a more Robust CLI configuration is to enforce the Builder, and L2 Client JWT are provided explicitly even if they are the same. In most configuration's it seems very likely they will not be the same, so silently inferring they are identical could cause silent bugs after startup. 

Let me know what you think

avalonche commented on src/server.rs:
```diff
@@ -356,7 +550,7 @@ where
             };
             if payload_status.is_invalid() {
-                error!(message = "builder payload was not valid", "url" = builder.url, "payload_status" = %payload_status.status, "payload_id" = %payload_id);
+                error!(message = "builder payload was not valid",  "payload_status" = %payload_status.status, "payload_id" = %payload_id);
```
missing ?self.builder_client.auth_socket?

0xKitsune commented:
I have added a set of `FORWARD_REQUESTS` and updated the `ProxyService` to implement the forwarding logic. 

```rust
  fn call(&mut self, req: HttpRequest<HttpBody>) -> Self::Future {
     // --snip--

            if MULTIPLEX_METHODS.iter().any(|&m| method.starts_with(m)) {
                if FORWARD_REQUESTS.iter().any(|&m| method.starts_with(m)) {
                    let builder_client = client.clone();
                    let builder_req =
                        HttpRequest::from_parts(parts.clone(), HttpBody::from(body_bytes.clone()));
                    let builder_method = method.clone();

                    tokio::spawn(async move {
                        let _ = forward_request(
                            builder_client,
                            builder_req,
                            &builder_method,
                            builder_uri,
                            builder_auth,
                        )
                        .await;
                    });

                    let l2_req = HttpRequest::from_parts(parts, HttpBody::from(body_bytes));
                    info!(target: "proxy::call", message = "proxying request to rollup-boost server", ?method);
                    forward_request(client, l2_req, &method, l2_uri, None).await

       // --snip--
    }

```

I also removed the `EthApi`, `MinerApi` and `MinerApiExt` impls from `server.rs` in favor of the `forward_request` logic in the proxy. I will update tests, address all remaining comments and clean up the PR tomorrow for another round of review.

0xOsiris commented on src/proxy.rs:
```diff
@@ -74,48 +106,91 @@ where
+
+                    let l2_req = HttpRequest::from_parts(parts, HttpBody::from(body_bytes));
+                    info!(target: "proxy::call", message = "proxying request to rollup-boost server", ?method);
+                    forward_request(client, l2_req, &method, l2_uri, None).await
```
@0xKitsune Why are we forwarding to the L2 clients auth URI, and the builder RPC URI. This will fail to send all `miner_` requests to the L2 client currently as the miner API doesn't exist on the Auth API

0xOsiris replied:
I think we need to pass both the l2 Auth URI, and the l2 RPC URI into the proxy layer. 

`FORWARD_REQUESTS` can target the RPC URI. 

Requests not in `MULTIPLEX_METHODS` can target the Auth URI.
